File: gymApp_backend/models/member_model.js
const { Timestamp } = require("bson");
const mongoose = require("mongoose");
const Schema = mongoose.Schema;

// Helper function to calculate the planExpiryDate in IST
const calculatePlanExpiryDate = (startDate, membershipPeriod) => {
  let date = new Date(
    startDate.getFullYear(),
    startDate.getMonth() + membershipPeriod,
    startDate.getDate()
  );

  // Convert to IST
  date = new Date(date.getTime() + 5.5 * 60 * 60 * 1000);
  return date;
};

const calculateDueAmount = (actualAmount, paidAmount) => {
  const dueAmount = actualAmount - paidAmount;
  return dueAmount;
};

// Helper function to get current date in IST
const getCurrentDateIST = () => {
  const currentDate = new Date();
  return new Date(currentDate.getTime() + 5.5 * 60 * 60 * 1000);
};

const memberSchema = new Schema(
  {
    first_name: String,
    last_name: String,
    phone_num: Number,
    gender: String,
    medicalIssue: String,
    membership_Period: Number,
    profile_img: String,
    actual_amount: Number,
    paid_amount: Number,
    due_amount: {
      type: Number,
      default: function () {
        return calculateDueAmount(this.actual_amount, this.paid_amount);
      },
    },
    planStartDate: {
      type: Date,
      default: getCurrentDateIST,
    },
    planExpiryDate: {
      type: Date,
      default: function () {
        return calculatePlanExpiryDate(
          this.planStartDate,
          this.membership_Period
        );
      },
    },
  },
  { timestamps: true }
);

// Pre-save middleware to adjust dates to IST
memberSchema.pre("save", function (next) {
  if (this.isModified("planStartDate")) {
    this.planStartDate = new Date(
      this.planStartDate.getTime() + 5.5 * 60 * 60 * 1000
    );
  }
  if (this.isModified("planExpiryDate")) {
    this.planExpiryDate = new Date(
      this.planExpiryDate.getTime() + 5.5 * 60 * 60 * 1000
    );
  }
  next();
});

// Virtual for daysRemaining
memberSchema.virtual("daysRemaining").get(function () {
  const today = new Date();
  const expiryDate = this.planExpiryDate;
  const timeDiff = expiryDate.getTime() - today.getTime();
  return Math.ceil(timeDiff / (1000 * 3600 * 24));
});

// Virtual for expired
memberSchema.virtual("expired").get(function () {
  return this.daysRemaining <= 0;
});

// Ensure virtual fields are serialized
memberSchema.set("toJSON", { virtuals: true });
memberSchema.set("toObject", { virtuals: true });

const Member = mongoose.model("Member", memberSchema);

module.exports = Member;

File: gym_app/lib/models/member_model.dart
class MemberModel {
  String? id;
  String? firstName;
  String? lastName;
  int? phoneNum;
  String? gender;
  String? profileImg;
  String? medicalIssue;
  int? membershipPeriod;
  int? actualAmount; // mentioned as "package" somewhere
  int? paidAmount;
  int? dueAmount;
  int? paidDueAmount;
  DateTime? planStartDate;
  DateTime? planExpiryDate;
  int? daysRemaining;
  bool? expired;

  MemberModel({
    this.id,
    this.firstName,
    this.lastName,
    this.phoneNum,
    this.gender,
    this.profileImg,
    this.medicalIssue,
    this.membershipPeriod,
    this.actualAmount,
    this.paidAmount,
    this.paidDueAmount,
    //auto calc
    this.dueAmount,
    this.planStartDate,
    this.planExpiryDate,
    this.daysRemaining,
    this.expired,
  });

  //Converts JSON object into member model
  factory MemberModel.fromJson(Map<String, dynamic> json) {
    return MemberModel(
      id: json['_id'],
      profileImg: json['profile_img'],
      firstName: json['first_name'],
      lastName: json['last_name'],
      phoneNum: json['phone_num'],
      gender: json['gender'],
      medicalIssue: json['medicalIssue'],
      membershipPeriod: json['membership_Period'],
      actualAmount: json['actual_amount'],
      paidAmount: json['paid_amount'],
      dueAmount: json['due_amount'],
      // paidDueAmount: json['paidDueAmount'],
      planStartDate: DateTime.parse(json['planStartDate']),
      planExpiryDate: DateTime.parse(json['planExpiryDate']),
      daysRemaining: json['daysRemaining'],
      expired: json['expired'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'membershipPeriod': membershipPeriod,
      'actualAmount': actualAmount,
      'paidAmount': paidAmount,
      'dueAmount': dueAmount,
      'paidDueAmount': paidDueAmount,
    };
  }
}
